### PR TITLE
Make Link become a template which takes a Clock parameter

### PR DESCRIPTION
**Pull request by abique**
This allows to use a custom clock, for example the one coming from the audio interface or simply the realtime clock.

I've made a similar request in the past and it was discarded, but I'd like you to consider it again.
In Bitwig we don't use the MONOTONIC_RAW clock, we use the realtime one, because this is the same clock being used by alsa, by pthread_*_timedwait() functions, by semaphore wait functions, etc...

You can't use the monotonic clock with every apis, which makes the realtime one the only clock that work across all the system interfaces.

I do not mean that the realtime clock should be the default, but it should be an option to use it.

Thank youl,
Alex

### CODE REVIEW

**Comment by abique** [comment]
To give you some more context, in BitwigStudio we want to use the same clock to timestamp MIDI input events, to schedule MIDI output events, to get timing from the audio interface and for Ableton Link.

In the middle of everything we use a bunch of synchronization primitives, with the one from pthread we can (most of the time) configure which clock to use (`CLOCK_REALTIME, CLOCK_MONOTONIC{,_RAW}, CLOCK_BOOTTIME, ...`); but there is one syscall that forces us to use `CLOCK_REALTIME` which is
`int sem_timedwait(sem_t *sem, const struct timespec *abs_timeout);`

In theory the ideal clock might be `CLOCK_MONOTONIC_RAW`, yet it could be `CLOCK_TAI` instead (first time I read about it) or `CLOCK_BOOTTIME`.

My point is that it is worth to allow the application to choose the clock.

As a reminder, I will copy paste the description of all the system clocks that we have today.

```
       Sufficiently recent versions of glibc and the Linux kernel support the
       following clocks:

       CLOCK_REALTIME
              A settable system-wide clock that measures  real  (i.e.,  wall-
              clock)  time.   Setting  this clock requires appropriate privi‐
              leges.  This clock is affected by discontinuous  jumps  in  the
              system time (e.g., if the system administrator manually changes
              the clock), and by the incremental adjustments performed by ad‐
              jtime(3) and NTP.

       CLOCK_REALTIME_ALARM (since Linux 3.0; Linux-specific)
              Like CLOCK_REALTIME, but not settable.  See timer_create(2) for
              further details.

       CLOCK_REALTIME_COARSE (since Linux 2.6.32; Linux-specific)
              A faster but less  precise  version  of  CLOCK_REALTIME.   This
              clock  is  not  settable.  Use when you need very fast, but not
              fine-grained timestamps.   Requires  per-architecture  support,
              and  probably  also  architecture  support for this flag in the
              vdso(7).

       CLOCK_TAI (since Linux 3.10; Linux-specific)
              A nonsettable system-wide clock derived  from  wall-clock  time
              but ignoring leap seconds.  This clock does not experience dis‐
              continuities and backwards jumps caused by NTP  inserting  leap
              seconds as CLOCK_REALTIME does.

              The acronym TAI refers to International Atomic Time.

       CLOCK_MONOTONIC
              A  nonsettable system-wide clock that represents monotonic time
              since—as described by  POSIX—"some  unspecified  point  in  the
              past".   On Linux, that point corresponds to the number of sec‐
              onds that the system has been running since it was booted.

              The CLOCK_MONOTONIC clock  is  not  affected  by  discontinuous
              jumps  in  the  system  time (e.g., if the system administrator
              manually changes the clock), but is affected by the incremental
              adjustments  performed  by adjtime(3) and NTP.  This clock does
              not count time that the system is suspended.   All  CLOCK_MONO‐
              TONIC  variants guarantee that the time returned by consecutive
              calls will not go backwards, but successive calls may—depending
              on  the architecture—return identical (not-increased) time val‐
              ues.

       CLOCK_MONOTONIC_COARSE (since Linux 2.6.32; Linux-specific)
              A faster but less precise version of CLOCK_MONOTONIC.  Use when
              you  need very fast, but not fine-grained timestamps.  Requires
              per-architecture support, and probably also  architecture  sup‐
              port for this flag in the vdso(7).

       CLOCK_MONOTONIC_RAW (since Linux 2.6.28; Linux-specific)
              Similar  to CLOCK_MONOTONIC, but provides access to a raw hard‐
              ware-based time that is not subject to NTP adjustments  or  the
              incremental  adjustments  performed  by adjtime(3).  This clock
              does not count time that the system is suspended.

       CLOCK_BOOTTIME (since Linux 2.6.39; Linux-specific)
              A  nonsettable  system-wide  clock   that   is   identical   to
              CLOCK_MONOTONIC, except that it also includes any time that the
              system is suspended.  This allows applications to  get  a  sus‐
              pend-aware monotonic clock without having to deal with the com‐
              plications of CLOCK_REALTIME, which may have discontinuities if
              the time is changed using settimeofday(2) or similar.

       CLOCK_BOOTTIME_ALARM (since Linux 3.0; Linux-specific)
              Like CLOCK_BOOTTIME.  See timer_create(2) for further details.

       CLOCK_PROCESS_CPUTIME_ID (since Linux 2.6.12)
              This is a clock that measures CPU time consumed by this process
              (i.e., CPU time consumed by all threads in  the  process).   On
              Linux, this clock is not settable.

       CLOCK_THREAD_CPUTIME_ID (since Linux 2.6.12)
              This is a clock that measures CPU time consumed by this thread.
              On Linux, this clock is not settable.

       Linux also implements dynamic clock instances as described below.

   Dynamic clocks
       In addition to the  hard-coded  System-V  style  clock  IDs  described
       above, Linux also supports POSIX clock operations on certain character
       devices.  Such devices are called "dynamic" clocks, and are  supported
       since Linux 2.6.39.

       Using  the  appropriate macros, open file descriptors may be converted
       into clock IDs and passed  to  clock_gettime(),  clock_settime(),  and
       clock_adjtime(2).   The  following example shows how to convert a file
       descriptor into a dynamic clock ID.

           #define CLOCKFD 3
           #define FD_TO_CLOCKID(fd)   ((~(clockid_t) (fd) << 3) | CLOCKFD)
           #define CLOCKID_TO_FD(clk)  ((unsigned int) ~((clk) >> 3))

           struct timeval tv;
           clockid_t clkid;
           int fd;

           fd = open("/dev/ptp0", O_RDWR);
           clkid = FD_TO_CLOCKID(fd);
           clock_gettime(clkid, &tv);
```

**Comment by abique** [comment]
I've rebased and squashed the commits.

**Comment by cdi-ableton** [comment]
Thanks for the PR. The changes make sense. I'm not too happy with the name `BasicLink` though and the documentation needs an update. I'll have a chat with some folks when they're back from vacation and will get back to you.

**Comment by abique** [comment]
Thank you very much.
I can also join in a google hangout meeting if you want to chat.

Regarding `BasicLink` it is a little bit like `std::basic_string`. I'm open to suggestions.

**Comment by abique** [comment]
Hi,
Did you think of a better name than BasicLink?
Regards,
Alex

**Comment by cdi-ableton** [comment]
Hej @abique. Sorry, this took us some time. We came to the conclusion that you already picked the most suitable name. Before merging we'd like to get the updated documentation ready so that we can push a follow-up commit right after merging. We'll ping you once we're ready.

**Comment by abique** [comment]
Thank you for reviewing.
I'll add the documentation to the change today I hope.

**Comment by cdi-ableton** [comment]
You don't have to do the documentation. We will do that and commit it as a follow-up to your changes. Your PR will be merged as soon as the documentation is ready.

**Comment by abique** [comment]
Ah OK, thank you very much!

**Comment by abique** [comment]
Hi,

Did you make progress with the documentation?

Regards,
Alexandre

File: include/ableton/Link.hpp
Comment (cdi-ableton) [inline include/ableton/Link.hpp:385]: In our setup we forward declare `Link`. This turned out to be complicated using a type alias. Therefore I suggest sticking with a `Link` class, as it is more in line with what we had before:
```
class Link : public BasicLink<link::platform::Clock>
{
public:
  Link(double bpm)
    : BasicLink(bpm)
  {
  }
};
```

File: include/ableton/Link.ipp
Comment (cdi-ableton) [inline include/ableton/Link.ipp:22]: This is not neccessary since we include `Link.ipp` in `Link.hpp`.

File: include/ableton/link/Controller.hpp
Comment (cdi-ableton) [inline include/ableton/link/Controller.hpp:122]: We should not rename `Clock` to stay consistent with the rest of the code base.

File: include/ableton/Link.hpp
Comment (cdi-ableton) [inline include/ableton/Link.hpp:369]: Please delete the now unused type alias in https://github.com/Ableton/link/blob/aa477048017a4d4fa28922bab33da3bdb3c2dabf/include/ableton/platforms/Config.hpp#L79

**Review by cdi-ableton — CHANGES_REQUESTED** [review]
The documentation is ready and I will add it after this PR is merged. I have some change requests before though.

File: include/ableton/Link.ipp
Comment (abique) [reply include/ableton/Link.ipp:22]: OK

File: include/ableton/link/Controller.hpp
Comment (abique) [reply include/ableton/link/Controller.hpp:122]: But then the code becomes confusing because there is a Clock declaration somewhere for the default clock isn't it?

File: include/ableton/Link.hpp
Comment (abique) [reply include/ableton/Link.hpp:369]: OK

File: include/ableton/link/Controller.hpp
Comment (cdi-ableton) [reply include/ableton/link/Controller.hpp:122]: I'd like to stick to `Clock`. From the `Controller`'s perspective nothing really changed regarding the default clock. There is only the type alias in `Config.hpp` and then the clock is injected into the `Controller` and the rest of Link. The difference is that the `Controller` type alias is now declared in `Link.hpp` instead of `Config.hpp`.

File: include/ableton/Link.hpp
Comment (abique) [reply include/ableton/Link.hpp]: Done.

File: include/ableton/Link.ipp
Comment (abique) [reply include/ableton/Link.ipp]: Done

File: include/ableton/link/Controller.hpp
Comment (abique) [reply include/ableton/link/Controller.hpp]: Done.

File: include/ableton/Link.hpp
Comment (abique) [reply include/ableton/Link.hpp:369]: Done.

**Comment by abique** [comment]
I don't think that the compile error from AppVeyor is related to my branch:

```
C:\projects\link\include\ableton/link/GhostXForm.hpp(36,55): error C2220: the following warning is treated as an error (compiling source file C:\projects\link\examples\linkhut\main.cpp) [C:\projects\link\build\examples\LinkHutSilent.vcxproj]
C:\projects\link\include\ableton/link/GhostXForm.hpp(36,55): warning C5219: implicit conversion from '_Rep' to 'const double', possible loss of data [C:\projects\link\build\examples\LinkHutSilent.vcxproj]
          with
          [
              _Rep=__int64
          ] (compiling source file C:\projects\link\examples\linkhut\main.cpp)
C:\projects\link\include\ableton/link/GhostXForm.hpp(36,55): message : to simplify migration, consider the temporary use of /Wv:18 flag with the version of the compiler with which you used to build without warnings (compiling source file C:\projects\link\examples\linkhut\main.cpp) [C:\projects\link\build\examples\LinkHutSilent.vcxproj]
C:\projects\link\include\ableton/link/GhostXForm.hpp(41,62): warning C5219: implicit conversion from '_Rep' to 'const double', possible loss of data [C:\projects\link\build\examples\LinkHutSilent.vcxproj]
          with
          [
              _Rep=__int64
          ] (compiling source file C:\projects\link\examples\linkhut\main.cpp)
C:\projects\link\include\ableton/link/GhostXForm.hpp(41,62): message : to simplify migration, consider the temporary use of /Wv:18 flag with the version of the compiler with which you used to build without warnings (compiling source file C:\projects\link\examples\linkhut\main.cpp) [C:\projects\link\build\examples\LinkHutSilent.vcxproj]
```

**Comment by cdi-ableton** [comment]
Ah. That's due to a Visual Studio 2019 update. It's fixed now. Please rebase your branch onto `master`.

**Comment by abique** [comment]
Done.

**Review by cdi-ableton — APPROVED** [review]

**Comment by abique** [comment]
Thank you! :+1:

**Comment by cdi-ableton** [comment]
Dito!

**Pull request merged**
